### PR TITLE
Report fewer spurious cyclic errors

### LINKED ISSUE #8252: spurious "cyclic aliasing" error
```scala
  import scala.language.higherKinds
  type Id[A] = A
  def foo[F[_], G[_], A](a: F[G[A]]): F[G[A]] = { println("what?! " + a); a }
  def oops(): Unit = {
    foo[Id, Id, Int](1) // compiles fine
  }
  oops() // prints "what?! 1"
  def expected(): Unit = {
    // error: cyclic aliasing or subtyping involving type Id
    val kaboom = foo[Id, Id, Int](1) 
  }
```

### LINKED ISSUE #9568: inappropriate cyclic aliasing
I catch strange type checker behavior that produce type error from nowhere.

```scala
object Cyclic {
  class Node[T]() {
    type Self = T
  }

  val nodeA = new Node[Int]()
  val nodeB = new NodeB(a => a)
  val nodeC = new NodeC(a => a)
  /*
  val nodeB = new NodeB(a => a + 1)
  val nodeC = new NodeC(a => a + 1)
  */
  val nodeD = new NodeD( (b,c) => b + c )

  class NodeB[T](fun : Function[nodeA.Self, T]) extends Node[T]
  class NodeC[T](fun : Function[nodeA.Self, T]) extends Node[T]
  class NodeD[T](fun : Function2[nodeB.Self, nodeC.Self, T]) extends Node[T]
}
```

The snipped produce following error:

```scala
CyclicAliasing.scala:98: error: cyclic aliasing or subtyping involving type Self
  val nodeD = new NodeD( (b,c) => b + c )
                          ^
CyclicAliasing.scala:98: error: cyclic aliasing or subtyping involving type Self
  val nodeD = new NodeD( (b,c) => b + c )
                            ^
two errors found
```

I could not reduce the example further. Small changes miraculously cure the code and it compiles. I gave two examples as comments. Small change *i => i*  with *i => i + 1* make type checker happy.

### LINKED ISSUE #10669: spurious "cyclic aliasing" error (2)
```scala
abstract class CharSet[P] {
  type Type <: P
}

object LetterOrDigit extends CharSet[Char]
object Digit extends CharSet[LetterOrDigit.Type]

type D = Digit.Type // error: cyclic aliasing or subtyping involving type Type
```

Possibly related to #8252 

---
Without parent trait:
```scala
object LetterOrDigit {
  type Type <: Char
}
object Digit {
  type Type <: LetterOrDigit.Type
}
type D = Digit.Type // works fine
```

Another workaround:
```scala
trait CharSet[P] {
  type Type <: P
}
object LetterOrDigit extends { type Type <: Char } with CharSet[Char]
object Digit extends { type Type <: LetterOrDigit.Type } with CharSet[LetterOrDigit.Type]
type D = Digit.Type // works fine
```

### LINKED ISSUE #12622: new "cyclic aliasing or subtype" error in 2.13
## Reproduction steps

```scala
object repro {
  trait ScenarioParam {
    type Builder <: Type
  }

  trait ScenarioParamBuilder

  trait Type {
    type Builder <: ScenarioParamBuilder
  }

  trait Types[H <: ScenarioParam, T <: Type] extends Type {
    type Builder = H#Builder with T#Builder
  }

  trait Nil extends Type {
    type Builder = ScenarioParamBuilder
  }

  trait ScenarioTarget {
    type FilterParam <: Type
  }

  class P1 extends ScenarioParam
  class P2 extends ScenarioParam

  object someTarget extends ScenarioTarget {
    type FilterParam = Types[P1, Types[P2, Nil]]
  }

  class WhereClauseBuilder1[T <: ScenarioTarget] {
    type FilterBuilderType = T#FilterParam#Builder
    def m1(f: FilterBuilderType => Any): Any = null
    def m2(f: T#FilterParam#Builder => Any): Any = null
  }
  
  // triggers cyclic aliasing error on 2.13 only
  (null: WhereClauseBuilder1[someTarget.type]).m1(x => null)

  // triggers the cyclic aliasing error on 2.12 and 2.13
  // this corresponds to the desugaring of the above with a with PR #5999
  val stabilizer: WhereClauseBuilder1[someTarget.type] = null
  stabilizer.m1(x => null)
  
  // also triggers the cyclic aliasing error on 2.12 and 2.13
  (null: WhereClauseBuilder1[someTarget.type]).m2(x => null)

}
```

```
$ scalac  --scala-version 2.13.8 -d /tmp test/files/pos/stabilizer-cyclic.scala
test/files/pos/stabilizer-cyclic.scala:38: error: cyclic aliasing or subtyping involving type Builder
  (null: WhereClauseBuilder1[someTarget.type]).m1(x => null)
                                                  ^
test/files/pos/stabilizer-cyclic.scala:43: error: cyclic aliasing or subtyping involving type Builder
  stabilizer.m1(x => null)
                ^
test/files/pos/stabilizer-cyclic.scala:46: error: cyclic aliasing or subtyping involving type Builder
  (null: WhereClauseBuilder1[someTarget.type]).m2(x => null)
                                                  ^
3 errors

$ scalac  --scala-version 2.12.15 -d /tmp test/files/pos/stabilizer-cyclic.scala
test/files/pos/stabilizer-cyclic.scala:43: error: cyclic aliasing or subtyping involving type Builder
  stabilizer.m1(x => null)
                ^
test/files/pos/stabilizer-cyclic.scala:46: error: cyclic aliasing or subtyping involving type Builder
  (null: WhereClauseBuilder1[someTarget.type]).m2(x => null)
                                                  ^
```

The relevant change in 2.13 appears to be https://github.com/scala/scala/pull/5999 (stabilising vals)
  
`scalac -Yrecursion 1` avoids all errors in both 2.12 and 2.13.

### LINKED ISSUE #12814: `error during expansion of this match (this is a scalac bug)` after upgrade from 2.13.10 to 2.13.11
The following code causes compilation error under 2.13.11

File `bug.sc`

```scala
//> using scala "2.13.11"

sealed trait Common
sealed trait One extends Common
sealed trait Two extends Common


trait Module[C <: Common] {
  val name: String
  type Narrow = C
  def narrow: PartialFunction[Common, C]
}

object ModuleA extends Module[One] {
  val name = "A"
  val narrow: PartialFunction[Common, Narrow] = {
    case cc: Narrow => cc
  }
}

object ModuleB extends Module[ModuleA.Narrow] {
  val name = "B"
  val narrow: PartialFunction[Common, Narrow] = {
    case cc: Narrow => cc
  }
}

object ModuleC extends Module[Two] {
  val name = "C"
  val narrow: PartialFunction[Common, Narrow] = {
    case cc: Narrow => cc
  }
}

object ModuleD extends Module[One with Two] {
  val name = "D"
  val narrow: PartialFunction[Common, Narrow] = {
    case cc: Narrow => cc
  }
}

val one    = new One {}
val two    = new Two {}
val oneTwo = new One with Two {}

Seq(ModuleA, ModuleB, ModuleC, ModuleD).foreach { module =>
  println(s"${module.name} at One    = ${module.narrow.isDefinedAt(one)}")
  println(s"${module.name} at Two    = ${module.narrow.isDefinedAt(two)}")
  println(s"${module.name} at OneTwo = ${module.narrow.isDefinedAt(oneTwo)}")
  println("-" * 10)
}

```

Run with:

```bash
scala-cli bug.sc
```

Compiler error:

```
Compiling project (Scala 2.13.11, JVM)
[error] ./bug.sc:23:49
[error] error during expansion of this match (this is a scalac bug).
[error] The underlying error was: cyclic aliasing or subtyping involving type Narrow
[error]   val narrow: PartialFunction[Common, Narrow] = {
[error]                                                 ^
Error compiling project (Scala 2.13.11, JVM)
Compilation failed
```

It was not the case under 2.13.10, where code compiles successfully.

Run under 2.13.10, change the first line to

```scala
//> using scala "2.13.10"
```

and run as before

```bash
scala-cli bug.sc
```

### PR DESCRIPTION
**Pull request by lrytz**
Fixes https://github.com/scala/bug/issues/12814, fixes https://github.com/scala/bug/issues/12622, fixes https://github.com/scala/bug/issues/10669, fixes https://github.com/scala/bug/issues/9568, fixes https://github.com/scala/bug/issues/8252.

Co-authored by @retronym.

When checking type params / abstract types for cyclic subtyping, don't use `symbol.lock` to detect cycles.

The same symbol may occur multiple times and expand to a different type, for example `a.Id[a.Id[Int]]`.

Seen types need to be compared with `==` (not `=:=`), otherwise `a.Id[Int]` is matches `a.Id[a.Id[Int]]`.

I tried using `symbol.lock` in a first step and only adding types to the stack from the second iteration, but some of the new test cases then still fail to compile because the same symbol lock is also used for type completers (detecting cyclic references).

An issue was that diverging types were no longer reported as cyclic, as in neg/t510.scala. I added a recursion depth limit for this case.

```
abstract class C { type T }
abstract class D[S <: C](val c: S) extends C {
  type T <: c.T
}
abstract class E(e: E) extends D[E](e)
object Test {
  def f(e: E): Unit = {
    def g(t: e.T): Unit = ()
    ()
  }
}
```

`e.T` keeps expanding to `e.c.T` to `e.c.c.T` and so on.

Note that removing `object Test` from the example and compiling only the definitions runs into a stack overflow in a later phase, so the compiler is vulnerable to diverging types in other places.

For the record, in Scala 3 catches stack overflows and reports "Recursion limit exceeded" errors.

### CODE REVIEW

**Comment by lrytz** [comment]
Dotty does not count / limit the recursion depth, the nicer error message is crafted when catching a `StackOverflowError`. For example, see uses of `handleRecursive` in https://github.com/lampepfl/dotty/blob/main/tests/pos-with-compiler-cc/dotc/core/Types.scala. I think it's fine to let the stack overflow instead of trying to guard against that, diverging types shouldn't happen often.

In this light, this PR might be the right tradeoff. Some code with diverging types that previously triggered to a `cyclic aliasing` error now leads to a stack overflow. I'll see if we can catch those and turn them into error reports like Scala 3. In exchange we get rid of spurious `cyclic aliasing` errors that prevents valid code from compiling.

**Comment by retronym** [comment]
Catching SOE is a bit risky because the exception handler can itself trigger another SOE before when it is doing its logging/reporting/throwing of a "better" exception. But perhaps for a compiler its okay.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -425,19 +428,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       }
     }
 
-    def checkNonCyclic(pos: Position, tp: Type, lockedSym: Symbol): Boolean = try {
-      if (!lockedSym.lock(CyclicReferenceError(pos, tp, lockedSym))) false
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:429]: `lock` always succeeded. This overload of `checkNonCyclic` is invoked right after `checkNotLocked`.

**Comment by lrytz** [comment]
> Catching SOE is a bit risky

Added a depth limit for `checkNonCyclic`.

Updated the PR description, ready for review @som-snytt @retronym.

**Review by retronym — APPROVED** [review]
Thank you for reviving this. Having a separate recursion counter seems like a good approach.

On the way to a failure, we have a `O(N^2)` performance with the `tps.contains(tp)`.  I think the limit of the list length is small enough to not worry about this.

It is easy to write code that that unwittngly teeters on the edge of the previous implementation's cliff of spuriousness. If that edge slightly shifts, as happened with Scala 2.13, it can be nigh on impossible to refactor the code in such a a way to retreat to safety.

**Comment by SethTisue** [comment]
community build kicked off — I'll report back if there's a problem, otherwise no news is good news

**Pull request merged**